### DISCUSSION TITLE
How to add a Script mapper in keycloak

### DISCUSSION THREAD

**Discussion by anishagoyal0504** [keycloak/keycloak#32639] [Q&A]
Hi Team,
I an trying to integrate keycloak as a OIDC with VMware Vcenter server. I came across one of the blog shared by someone how to do this?

They said I need to create few mappers and one of them is script mapper so wondering how do I create script mapper, In new UI of keycloak I cannot see this mapper?

Do I have to enable some sort of Javascript plugin if yes, how to do it?
Can someone guide me please

**Comment by bonnm** [comment]
How to add script mappers to Keycloak:
https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers
https://www.keycloak.org/docs/latest/server_development/index.html#openid-connect-protocol-mapper

I guess the relevant script mapper needed for vcenter is one to overwrite the sub claim with a user value different from the UUID that Keycloak uses by default as sub claim.

The code within the script mapper then is something like this
`token.setSubject(user.PROPERTY_NEEDED_BY_VCENTER);`
or
`token.setSubject(user.ATTRIBUTE_NEEDED_BY_VCENTER[0]);`

**Comment by InfoSec812** [comment]
I just recently worked this out for my SAML client.

I first had to create a script file `full-name-mapper.js` with the following contents:

```javascript
var user = keycloakSession.getContext().getUser();
var fullName = user.getFirstName() + " " + user.getLastName()
exports = fullName
```

Next, I created a `META-INF/keycloak-scripts.json` file like this:

```json
{
    "saml-mappers": [
        {
            "name": "Full Name Mapper",
            "fileName": "full-name-mapper.js",
            "description": "Combine firstName and lastName in a single field"
        }
    ]
}
```

Then I had to create a JAR file like this:

```bash
jar -cvf mappers.jar full-name-mapper.js META-INF/keycloak-scripts.json
```

Finally, in my container build I have to copy the jar to `/opt/keycloak/providers` and set the environment variable `KC_FEATURES=preview` as shown below:

```Dockerfile
FROM quay.io/keycloak/keycloak:26.5 AS builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true
ENV KC_DB=postgres

# Configure a database vendor
ENV KC_PROXY_HEADERS=xforwarded
ENV KC_HTTP_ENABLED=true
ENV KC_DB=postgres
ENV KC_FEATURES=preview

COPY mappers.jar /opt/keycloak/providers/mappers.jar

WORKDIR /opt/keycloak
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:26.5
COPY --from=builder /opt/keycloak/ /opt/keycloak/
COPY . /opt/keycloak/themes/zanclus

# change these values to point to a running postgres instance
ENV KC_HTTP_ENABLED=true
ENV KC_DB=postgres
ENV KC_DB_URL=jdbc:postgresql://postgresql:5432/keycloak
ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
```

I then restarted my quadlet which is running Keycloak and the mapper is now listed:

<img width="881" height="384" alt="image" src="https://github.com/user-attachments/assets/e0e48e17-f7dc-48f5-8722-48176ce5bf74" />
<img width="856" height="501" alt="image" src="https://github.com/user-attachments/assets/fd3038b7-a75a-4c3a-9913-3cf41f66a446" />
